### PR TITLE
Implement feature AnimatedVisualPlayer

### PR DESCRIPTION
**Pull request by improbabilityphysics**
Full description of AVP(cited from simeon's words):

"The AnimatedVisualPlayer is a XAML FrameworkElement that hosts and controls animated Windows.UI.Composition content. This content is typically Lotties (see https://airbnb.design/lottie/) but can be anything that conforms to the IAnimatedVisualSource interface. The player is most typically used to display and control micro-animations within UI, for example an animated icon on a button.

To use the AnimatedVisualPlayer with Lottie files, see https://github.com/windows-toolkit/Lottie-Windows."

In this PR, I have made following changes:
1. merge AVP changes from winui_rs5
2. use USE_INSIDER_SDK for 19H1 preview because RS5 SDK has IClosable MIDLRT compiling error.
3. use USE_INTERNAL_SDK macro for massaging code because some API code is not exposed to insider preview yet.
4. renaming Deps to MUX related to namingspace
5. ASSERT macro modification.
6. use codegen for dependency properties of AVP.

### CODE REVIEW

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.h [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+
+    // IUIElement / IUIElementOverridesHelper
+    winrt::AutomationPeer OnCreateAutomationPeer();
+
+    static void ClearProperties();
+    static void EnsureProperties();
+
+    // Dependency properties
+    static winrt::DependencyProperty AutoPlayProperty() { return s_AutoPlayProperty; }
+    static winrt::DependencyProperty DiagnosticsProperty() { return s_DiagnosticsProperty; }
```
Comment (improbabilityphysics) [inline dev/AnimatedVisualPlayer/AnimatedVisualPlayer.h]: remove these dependency property related code and inherit this class from codegen property class for AnimatedViualPlayer.

File: dev/Generated/AnimatedVisualPlayer.properties.cpp [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+                winrt::name_of<winrt::AnimatedVisualPlayer>(),
+                false /* isAttached */,
+                ValueHelper<winrt::Stretch>::BoxedDefaultValue(),
+                nullptr);
+    }
+}
+
+void AnimatedVisualPlayerProperties::ClearProperties()
+{
+    s_AutoPlayProperty = nullptr;
```
Comment (improbabilityphysics) [inline dev/Generated/AnimatedVisualPlayer.properties.cpp]: @simeoncran , is it intentional not to clear s_AutoPlayProperty in the old code?

File: dev/Generated/AnimatedVisualPlayer.properties.cpp [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+GlobalDependencyProperty AnimatedVisualPlayerProperties::s_PlaybackRateProperty{ nullptr };
+GlobalDependencyProperty AnimatedVisualPlayerProperties::s_SourceProperty{ nullptr };
+GlobalDependencyProperty AnimatedVisualPlayerProperties::s_StretchProperty{ nullptr };
+
+AnimatedVisualPlayerProperties::AnimatedVisualPlayerProperties()
+{
+    EnsureProperties();
+}
+
+void AnimatedVisualPlayerProperties::EnsureProperties()
```
Comment (improbabilityphysics) [inline dev/Generated/AnimatedVisualPlayer.properties.cpp]: double check and see if all properties have values for working consistent to those in old code.

**Comment by improbabilityphysics** [comment]
@simeoncran @jevansaks @Microsoft/microsoft-ui-xaml-codereviews ping...

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.idl [hunk 4]
```diff
@@ -0,0 +1,83 @@
+﻿[WUXC_VERSION_INTERNAL]
```
Comment (jevansaks) [inline dev/AnimatedVisualPlayer/AnimatedVisualPlayer.idl]: I think these should be WUXC_VERSION_MUXONLY instead of "INTERNAL", since "INTERNAL" will tag them as experimental. But there was still an open discussion about some of these interfaces being not ship-ready. Did that close? Or will you move these to public later?

**Comment by jevansaks** [comment]
Can you fix the title of this PR to be the newly added feature with a full description about the feature? There's no context for this feature on the GitHub.

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.h [hunk 5]
```diff
@@ -0,0 +1,147 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#pragma once
+
+#include "pch.h"
+#include "common.h"
```
Comment (jevansaks) [inline dev/AnimatedVisualPlayer/AnimatedVisualPlayer.h]: pch.h and common.h should be included in the cpp file, not the headers.

File: dev/Interactions/ButtonInteraction/ButtonInteraction.idl [hunk 6]
```diff
@@ -17,7 +17,7 @@ runtimeclass ButtonInteractionInvokedEventArgs
 [WUXC_VERSION_PREVIEW]
 [webhosthidden]
 unsealed runtimeclass ButtonInteraction :
-#ifdef USE_INSIDER_SDK
+#ifdef USE_INTERNAL_SDK
```
Comment (jevansaks) [inline dev/Interactions/ButtonInteraction/ButtonInteraction.idl]: Please add build support for this new flag in the same way that UseInsiderSDK had it.

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.idl (hunk 4 above)
Comment (jevansaks) [reply dev/AnimatedVisualPlayer/AnimatedVisualPlayer.idl]: The PR is waiting signoff in the internal repo? Let's just move all the work to GitHub and abandon the other PRs/branches.

File: dev/Generated/AnimatedVisualPlayer.properties.cpp (hunk 2 above)
Comment (jevansaks) [reply dev/Generated/AnimatedVisualPlayer.properties.cpp]: I'm sure it was an oversight, that's why we generate this code. It was too easy to get wrong.

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.idl (hunk 4 above)
Comment (improbabilityphysics) [reply dev/AnimatedVisualPlayer/AnimatedVisualPlayer.idl]: Okay.

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp [hunk 7]
```diff
@@ -0,0 +1,974 @@
...
+        m_owner.IsPlaying(false);
+    }
+
+    // Allow the play to complete.
+    CompleteAwaits();
+}
+
+AnimatedVisualPlayer::AnimatedVisualPlayer()
+{
+    EnsureProperties();
```
Comment (FrankJWong) [inline dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp]: Add this macro:
__RP_Marker_ClassById(RuntimeProfiler::ProfId_AnimatedVisualPlayer);

In order to instrument class usage.  And include "RuntimeProfiler.h" to pull in the macro definition.

File: dev/Generated/AnimatedVisualPlayer.properties.cpp (hunk 3 above)
Comment (improbabilityphysics) [reply dev/Generated/AnimatedVisualPlayer.properties.cpp]: AutoPlay should be true in default. PlayBackRate should be 1 in default. Stretch should be ::Uniform in default. use MUX_DEFAULT_VALUE in AnimatedVisualPlayer.idl to change default value for each conflicted property value.

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.h [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+#include "AnimatedVisualPlayer.g.h"
+#include "AnimatedVisualPlayer.properties.h"
+#include <winrt\Windows.UI.Xaml.Controls.h>
```
[the remark this answers is not in the document]
Comment (improbabilityphysics) [reply dev/AnimatedVisualPlayer/AnimatedVisualPlayer.h]: Not really. Might these includes be in pch in more recent master commit?

[the remark this answers is not in the document]
Comment (improbabilityphysics) [reply dev/AnimatedVisualPlayer/AnimatedVisualPlayer.h]: Not really. Might these includes be in pch in more recent master commit?

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.h (hunk 5 above)
Comment (improbabilityphysics) [reply dev/AnimatedVisualPlayer/AnimatedVisualPlayer.h]: Not really. Might these includes be in pch in more recent master commit?

File: dev/Interactions/ButtonInteraction/ButtonInteraction.idl (hunk 6 above)
Comment (improbabilityphysics) [reply dev/Interactions/ButtonInteraction/ButtonInteraction.idl]: It looks like the changes are in environment.props and in Build.cmd. Working on it.

File: dev/Generated/AnimatedVisualPlayer.properties.cpp [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+{
+    if (!s_AutoPlayProperty)
+    {
+        s_AutoPlayProperty =
+            InitializeDependencyProperty(
+                L"AutoPlay",
+                winrt::name_of<bool>(),
+                winrt::name_of<winrt::AnimatedVisualPlayer>(),
+                false /* isAttached */,
```
Comment (improbabilityphysics) [inline dev/Generated/AnimatedVisualPlayer.properties.cpp]: Some properties are missing their callback functions. Please add them. Otherwise contents are not loaded as expected.

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (jevansaks) [reply dev/AnimatedVisualPlayer/AnimatedVisualPlayer.h]: CppWinRTIncludes.h has winrt\Windows.UI.Xaml.Controls.h which is included by pch.h.

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.h (hunk 5 above)
Comment (jevansaks) [reply dev/AnimatedVisualPlayer/AnimatedVisualPlayer.h]: I think you misunderstood my comment. The AnimatedVisualPlayer.h file should not have the lines #include "pch.h" or #include "common.h" -- those belong in the .cpp file.

File: test/MUXControlsTestApp/MSTest/MUXControlsTestApp.csproj [hunk 10]
```diff
@@ -36,6 +36,11 @@
       <SubType>Designer</SubType>
     </AppxManifest>
   </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="Win2D.uwp">
+      <Version>1.23.0</Version>
+    </PackageReference>
+  </ItemGroup>
```
Comment (jevansaks) [inline test/MUXControlsTestApp/MSTest/MUXControlsTestApp.csproj]: I suspect this will be a problem in the OS build, what took a dependency on Win2D?

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (improbabilityphysics) [reply dev/AnimatedVisualPlayer/AnimatedVisualPlayer.h]: Interesting, let me gibe it a try.

File: test/MUXControlsTestApp/MSTest/MUXControlsTestApp.csproj (hunk 10 above)
Comment (improbabilityphysics) [reply test/MUXControlsTestApp/MSTest/MUXControlsTestApp.csproj]: AnimatedVisualPlayer_TestUI LottieLogo.cs which is the content source for AnimatedVisualPlayer. This LottieLogo.cs has CanvasGeometry type which requires Win2D package. I recalled I have fixed this version to 1.21.0 or similar, not sure it is still 1.23.0. Let me double check and make sure it compiled in local.

Comment (jevansaks) [reply test/MUXControlsTestApp/MSTest/MUXControlsTestApp.csproj]: Is Win2d available in razzle? This code will be compiling in the OS repo.

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (improbabilityphysics) [reply dev/AnimatedVisualPlayer/AnimatedVisualPlayer.h]: removed #include <winrt\Windows.UI.Xaml.Controls.h>

File: test/MUXControlsTestApp/MSTest/MUXControlsTestApp.csproj (hunk 10 above)
Comment (improbabilityphysics) [reply test/MUXControlsTestApp/MSTest/MUXControlsTestApp.csproj]: Thanks to Simeon for his updating content source LottieLogo.cs which do not need to refer to Win2D.uwp package any more. This LottieLogo.cs has been updated and do not need to reference to Win2D any longer. And Win2D package is uninstalled.

File: MUXControls.sln [hunk 11]
```diff
@@ -850,20 +864,25 @@ Global
 		dev\SwipeControl\SwipeControl.vcxitems*{fd63f767-bd98-45d9-8a63-3cd4a148d527}*SharedItemsImports = 9
 	EndGlobalSection
 	GlobalSection(SolutionConfigurationPlatforms) = preSolution
+		Debug_test|Any CPU = Debug_test|Any CPU
```
Comment (jevansaks) [inline MUXControls.sln]: We should not have an "Any CPU" solution configuration. Please undo these changes.

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp [hunk 12]
```diff
@@ -0,0 +1,988 @@
...
+        if (playbackRate < 0)
+        {
+            // Play from end to beginning if playing in reverse.
+            m_controller.Progress(1);
+        }
+
+        if (m_batch)
+        {
+            // Subscribe to the batch completed event.
+            m_batchCompletedToken = m_batch.Completed([this](winrt::IInspectable const&, winrt::CompositionBatchCompletedEventArgs const&)
```
Comment (jevansaks) [inline dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp]: capturing "this" in deferred lambdas is dangerous. What guarantees this object lives longer than the callback?

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp [hunk 13]
```diff
@@ -0,0 +1,988 @@
...
+            m_batch.End();
+        }
+
+        m_owner.IsPlaying(true);
+    }
+}
+
+bool AnimatedVisualPlayer::AnimationPlay::IsCurrentPlay()
+{
+    return &*(m_owner.m_nowPlaying) == this;
```
Comment (jevansaks) [inline dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp]: The "&*" syntax seems odd. Assuming that shared_ptr doesn't have the comparison overload you're looking for can you use the .get() instead?

```suggestion
    return m_owner.m_nowPlaying.get() == this;
```

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp [hunk 14]
```diff
@@ -0,0 +1,988 @@
...
+    {
+        // Disconnect from the update notifications of the old source.
+        oldDynamicSource.AnimatedVisualInvalidated(m_dynamicAnimatedVisualInvalidatedToken);
+        m_dynamicAnimatedVisualInvalidatedToken = { 0 };
+    }
+
+    if (auto newDynamicSource = newSource.try_as<winrt::IDynamicAnimatedVisualSource>())
+    {
+        // Connect to the update notifications of the new source.
+        m_dynamicAnimatedVisualInvalidatedToken = newDynamicSource.AnimatedVisualInvalidated([this](auto const& /*sender*/, auto const&)
```
Comment (jevansaks) [inline dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp]: I don't see anything that disconnects this callback when the AnimatedVisualPlayer gets destroyed. Can you use auto_revoke instead?

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayerAutomationPeer.cpp [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+#include "AnimatedVisualPlayerAutomationPeer.h"
+
+#include "AnimatedVisualPlayer.h"
+
+AnimatedVisualPlayerAutomationPeer::AnimatedVisualPlayerAutomationPeer(winrt::AnimatedVisualPlayer const& owner)
+    : ReferenceTracker(owner)
+{
+}
+
+hstring AnimatedVisualPlayerAutomationPeer::GetClassNameCore()
```
Comment (jevansaks) [inline dev/AnimatedVisualPlayer/AnimatedVisualPlayerAutomationPeer.cpp]: why override GetClassNameCore? I don't see other peers doing this.

File: dev/CommandBarFlyout/CommandBarFlyout.cpp [hunk 16]
```diff
@@ -12,7 +12,7 @@ CommandBarFlyout::CommandBarFlyout()
 {
     __RP_Marker_ClassById(RuntimeProfiler::ProfId_CommandBarFlyout);
 
-#ifdef USE_INSIDER_SDK
+#ifdef USE_INTERNAL_SDK
```
Comment (jevansaks) [inline dev/CommandBarFlyout/CommandBarFlyout.cpp]: These APIs should be in the 18312 SDK, can you please include this under the "USE_INSIDER_SDK" ifdef?

File: dev/CommandBarFlyout/CommandBarFlyoutCommandBar.cpp [hunk 17]
```diff
@@ -292,7 +292,7 @@ void CommandBarFlyoutCommandBar::UpdateVisualState(bool useTransitions)
             bool isConstrainedToRootBounds = true;
             auto controlBounds = TransformToVisual(nullptr).TransformBounds({ 0, 0, static_cast<float>(ActualWidth()), static_cast<float>(ActualHeight()) });
             
-#ifdef USE_INSIDER_SDK
+#ifdef USE_INTERNAL_SDK
```
Comment (jevansaks) [inline dev/CommandBarFlyout/CommandBarFlyoutCommandBar.cpp]: These APIs should be in the 18312 SDK, can you please include this under the "USE_INSIDER_SDK" ifdef?

File: dev/NavigationView/NavigationView.cpp [hunk 18]
```diff
@@ -280,7 +280,7 @@ void NavigationView::OnApplyTemplate()
         auto visual = winrt::ElementCompositionPreview::GetElementVisual(topNavOverflowButton);
         CreateAndAttachHeaderAnimation(visual);
 
-#ifdef USE_INSIDER_SDK
+#ifdef USE_INTERNAL_SDK
```
Comment (jevansaks) [inline dev/NavigationView/NavigationView.cpp]: These APIs should be in the 18312 SDK, can you please include this under the "USE_INSIDER_SDK" ifdef?

File: dev/dll/Microsoft.UI.Xaml.vcxproj.filters [hunk 19]
```diff
@@ -108,5 +108,15 @@
...
+    <Page Include="@(PageRequiringCustomCompilation)" />
+    <Page Include="@(PageRequiringCustomCompilation)" />
+    <Page Include="@(PageRequiringCustomCompilation)" />
+    <Page Include="@(PageRequiringCustomCompilation)" />
+    <Page Include="@(PageRequiringCustomCompilation)" />
+    <Page Include="@(PageRequiringCustomCompilation)" />
+    <Page Include="@(PageRequiringCustomCompilation)" />
+    <Page Include="@(PageRequiringCustomCompilation)" />
+    <Page Include="@(PageRequiringCustomCompilation)" />
+    <Page Include="@(PageRequiringCustomCompilation)" />
```
Comment (jevansaks) [inline dev/dll/Microsoft.UI.Xaml.vcxproj.filters]: please revert these changes (git checkout origin/master __this file__)

File: test/MUXControlsTestApp/TAEF/project.json [hunk 20]
```diff
@@ -3,7 +3,8 @@
     "Microsoft.Net.Native.Compiler": "2.2.1",
     "Microsoft.NETCore.UniversalWindowsPlatform": "5.2.2",
     "MUXCustomBuildTasks": "1.0.38",
-    "TAEF.Redist.Wlk": "10.31.180822002"
+    "TAEF.Redist.Wlk": "10.31.180822002",
+    "Win2D.uwp": "1.22.0"
```
Comment (jevansaks) [inline test/MUXControlsTestApp/TAEF/project.json]: I thought you updated the sample to not need Win2D?

Comment (improbabilityphysics) [reply test/MUXControlsTestApp/TAEF/project.json]: Yes, will fix this.

**Comment by licanhua** [comment]
>   <Target Name="GenerateRS4GenericXamlFile" DependsOnTargets="CategorizeSharedPages" BeforeTargets="$(GenerateXamlFileBeforeTargets)" Condition="$(BuildingWithBuildExe) != 'true'" Inputs="@(RS1StylePage);@(RS2StylePage);@(RS3StylePage);@(RS4StylePage)" Outputs="$(OutDir)rs4_generic.xaml;">

Please double check with Alan, I think RS4 and RS5 is impacted too.

---
Refers to: dev/dll/Microsoft.UI.Xaml.vcxproj:587 in cc89861. [](commit_id = cc89861b6c6bc7f935c3f22fe485fbb7bcdd5a0f, deletion_comment = False)

**Comment by licanhua** [comment]
>   <Target Name="GenerateRS3ThemeResourceFile" DependsOnTargets="CategorizeSharedPages" BeforeTargets="$(GenerateXamlFileBeforeTargets)" Condition="$(BuildingWithBuildExe) != 'true'" Inputs="@(RS1ThemeResourcePage);@(RS2ThemeResourcePage);@(RS3ThemeResourcePage)" Outputs="$(OutDir)rs3_themeresources.xaml">

I expect you made change to this file too

---
Refers to: dev/dll/Microsoft.UI.Xaml.vcxproj:670 in cc89861. [](commit_id = cc89861b6c6bc7f935c3f22fe485fbb7bcdd5a0f, deletion_comment = False)

File: dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp (hunk 12 above)
Comment (improbabilityphysics) [reply dev/AnimatedVisualPlayer/AnimatedVisualPlayer.cpp]: Simeon and I worked very hard to analyze and do reasoning on behind this complicated scene. And comments have been added in the code for this good question.

**Comment by jevansaks** [comment]
> The latest preview SDK is 18323.. this PR is still targeting 18312

We can retarget in a subsequent change -- until the SDK for this release is finalized there's not too much reason to keep moving insider SDKs. You can use the Install-WindowsSdkISO.ps1 in the repo to install the archived ones.
